src/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -0,0 +1,430 @@
+  return _.mapValues(migrations, (versions, prop) => {
+    const transforms = Object.entries(versions)
+      .map(([version, transform]) => ({
+        version,
+        transform: wrapWithTry(version, prop!, transform),
+      }))
+      .sort((a, b) => Semver.compare(a.version, b.version));
+    return {
+      latestVersion: _.last(transforms).version,
+      transforms,
+    };
+  });
+}
+
+/**
+ * Creates a function which migrates and validates any document that is passed to it.
+ */
+function buildDocumentTransform({
+  kibanaVersion,
+  migrations,
+  validateDoc,
+}: {
+  kibanaVersion: string;
+  migrations: ActiveMigrations;
+  validateDoc: ValidateDoc;
+}): TransformFn {
+  return function transformAndValidate(doc: SavedObjectDoc) {
+    assertCanTransform(doc, kibanaVersion);
+
+    const result = doc.migrationVersion
+      ? applyMigrations(doc, migrations)
+      : markAsUpToDate(doc, migrations);
+
+    validateDoc(result);
+
+    // In order to keep tests a bit more stable, we won't
+    // tack on an empy migrationVersion to docs that have
+    // no migrations defined.
+    if (_.isEmpty(result.migrationVersion)) {
+      delete result.migrationVersion;
+    }
+
+    return result;
+  };
+}
+
+function applyMigrations(doc: SavedObjectDoc, migrations: ActiveMigrations) {
+  while (true) {
+    const prop = nextUnmigratedProp(doc, migrations);
+    if (!prop) {
+      return doc;
+    }
+    doc = migrateProp(doc, prop, migrations);
+  }
+}
+
+/**
+ * Gets the doc's props, handling the special case of "type".
+ */
+function props(doc: SavedObjectDoc) {
+  return Object.keys(doc).concat(doc.type);
+}
+
+/**
+ * Looks up the prop version in a saved object document or in our latest migrations.
+ */
+function propVersion(doc: SavedObjectDoc | ActiveMigrations, prop: string) {
+  return (
+    (doc[prop] && doc[prop].latestVersion) ||
+    (doc.migrationVersion && (doc as any).migrationVersion[prop])
+  );
+}
+
+/**
+ * Sets the doc's migrationVersion to be the most recent version
+ */
+function markAsUpToDate(doc: SavedObjectDoc, migrations: ActiveMigrations) {
+  return {
+    ...doc,
+    migrationVersion: props(doc).reduce((acc, prop) => {
+      const version = propVersion(migrations, prop);
+      return version ? _.set(acc, prop, version) : acc;
+    }, {}),
+  };
+}
+
+/**
+ * If a specific transform function fails, this tacks on a bit of information
+ * about the document and transform that caused the failure.
+ */
+function wrapWithTry(version: string, prop: string, transform: TransformFn) {
+  return function tryTransformDoc(doc: SavedObjectDoc) {
+    try {
+      const result = transform(doc);
+
+      // A basic sanity check to help migration authors detect basic errors
+      // (e.g. forgetting to return the transformed doc)
+      if (!result || !result.type) {
+        throw new Error(`Invalid saved object returned from migration ${prop}:${version}.`);
```

Comment:
Thoughts on logging `result` as well? This might help up down the road with user issues instead of having them inspect the data which failed.

Reply:
I like the idea. this will mean we are basically converting *every* doc to JSON on *every* transform, which seems unideal. What about doing a debug log of the end-result of all of the transforms on any given document?

Reply:
To loop back around here, this would just be for the error case and I believe we exit after the first error.